blog-post/blog-post.md

```diff
@@ -0,0 +1,382 @@
+The language now exhibits the scars of its accumulated complexity and a type system 
+that has become a barrier to its own evolution. Foundational features like 
+generics (aka parametric polymorphism) remain out of reach, as integrating them 
+into the current architecture is both difficult and dangerously prone to 
+introduce critical flaws.
+
+Aiming to break this impasse, the Argot collective decided to embark on a 
+fundamental rebuild, which is named Core Solidity: a project to define a new, 
+foundational version of the language designed for security, clarity, and 
+future growth from the ground up.
+
+
+## What is Core Solidity?
+
+The design of Core Solidity took inspiration from well established concepts from 
+functional programming, with Haskell serving as a primary influence, while 
+incorporating rigorous safety guarantees inspired by modern systems programming 
+languages such as Rust. Although this architectural direction represents a 
+significant evolution, it keeps the domain-specific requirements of smart 
+contract development, preserving the established conventions and practical 
+utility of Classic Solidity where appropriate.
+
+The language's theoretical foundation comprises several formally-specified 
+features that provide enhanced type safety and expressivity. In what follows, we 
+provide an overview of such features. However, we warn the reader that the current 
+Core Solidity prototype uses a syntax which is similar, but not identical, to 
+Classic Solidity. Since our initial concern is to develop the new language type 
+system and its semantics, its surface syntax would probably change in the near 
+future.
+
+### Generics and Type Classes
+
+Core Solidity will introduce two new exciting abstraction mechanisms: generics and 
+type classes.
+
+Generics enable parametric polymorphism through type parameters, which allows the  
+implementation of algorithms and data structures that operate uniformly across 
+all types. As an example, we could define a polymorphic `identity` function:
+
+``` 
+forall T . function identity(x : T) -> T {
+  return x;
+}
+```
+
+While generic functions are interesting, most interesting operations are not defined 
+at all types. Overloading allows the definition of code which can operate in distinct 
+ways at different types. Type classes are the standard way of combining overloading and 
+parametric polymorphism (generics) in a systematic manner. A type class definition 
+declares the class name, its arguments and member functions type signature. As an 
+example, let's consider the task of defining addition over different types:
+
+```
+forall T . class T : Sum {
+  function sum (x : T, y : T) -> T;
+```
+
+Implementations for member functions for different types are provided by instance 
+definitions. As an example, let's consider the implementation of `Sum` for 
+`word` type which, internally, uses Yul assembly to perform addition. 
```

Comment:
I think a large part of our audience will be unfamiliar with the term "instance" in this context. Perhaps we could add an extra sentence explaining that it's simply a specialization of the class?

Maybe even that it's the equivalent of `impl` in Rust. In the same vein we could say that a type class is the equivalent of `trait`.

Reply:
The `word` type is also something could use a short explanation. We don't have it in Classic Solidity and I don't think people will intuitively grasp what it is and how it relates to the integer types.

Reply:
> I think a large part of our audience will be unfamiliar with the term "instance" in this context. Perhaps we could add an extra sentence explaining that it's simply a specialization of a type class?
> 
> Maybe even that it's the equivalent of `impl` in Rust. In the same vein we could say that a type class is the equivalent of `trait`.

We are planning on purging these haskellisms entirely and probably just going with `trait` / `impl` since it's more widely used and less overloaded than `typeclass`. I don't think we will have time to do that before shipping this blog though, so will just for now have to live with the mess and make as clear as possible in the post that traits are typeclasses and instances are impls.